components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/DefaultServiceURLBuilder.java

```diff
@@ -93,29 +94,37 @@ public ServiceURL build() throws URLBuilderException {
         String proxyContextPath = ServerConfiguration.getInstance().getFirstProperty(PROXY_CONTEXT_PATH);
         String resolvedFragment = buildFragment(fragment, fragmentParams);
         String urlPath = getResolvedUrlPath(tenantDomain);
-
+        String relativePublicUrl = fetchRelativePublicUrl(proxyContextPath, urlPath, resolvedFragment);
+        String relativeInternalUrl = fetchRelativeInternalUrl(urlPath, resolvedFragment);
+        String absoluteInternalUrl = fetchAbsoluteInternalUrl(protocol, internalHostName, transportPort,
+                relativeInternalUrl);
+        String absolutePublicUrlWithoutURLPath = fetchAbsolutePublicUrlWithoutURLPath(protocol, proxyHostName,
+                proxyPort);
         if (StringUtils.isNotBlank(urlPath)) {
             if (authenticationEndpointHostName != null && authenticationEndpointPath != null &&
                     urlPath.contains(authenticationEndpointPath)) {
-                return new ServiceURLImpl(protocol, authenticationEndpointHostName, internalHostName, proxyPort,
-                        transportPort, tenantDomain, proxyContextPath, urlPath, parameters, resolvedFragment);
+                absolutePublicUrlWithoutURLPath = fetchAbsolutePublicUrlWithoutURLPath(protocol,
+                        authenticationEndpointHostName, proxyPort);
             }
             if (recoveryEndpointHostName != null && recoveryEndpointPath != null &&
                     urlPath.contains(recoveryEndpointPath)) {
-                return new ServiceURLImpl(protocol, recoveryEndpointHostName, internalHostName, proxyPort,
-                        transportPort, tenantDomain, proxyContextPath, urlPath, parameters, resolvedFragment);
+                absolutePublicUrlWithoutURLPath = fetchAbsolutePublicUrlWithoutURLPath(protocol,
+                        recoveryEndpointHostName, proxyPort);
             }
         }
+        String absolutePublicURL = fetchAbsolutePublicUrl(absolutePublicUrlWithoutURLPath, relativePublicUrl);
         return new ServiceURLImpl(protocol, proxyHostName, internalHostName, proxyPort, transportPort, tenantDomain,
-                proxyContextPath, urlPath, parameters, resolvedFragment);
+                proxyContextPath, urlPath, parameters, resolvedFragment, absolutePublicURL,
+                absoluteInternalUrl, relativePublicUrl, relativeInternalUrl, absolutePublicUrlWithoutURLPath);
     }
 
-    private String getResolvedUrlPath(String tenantDomain) {
+    protected String getResolvedUrlPath(String tenantDomain) {
 
         String resolvedUrlContext = buildUrlPath(urlPaths);
         StringBuilder resolvedUrlStringBuilder = new StringBuilder();
 
-        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
+        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled() &&
+                (!resolvedUrlContext.startsWith("t/") || !resolvedUrlContext.startsWith("/t/"))) {
```

Comment:
IIRC buildUrlPath(urlPaths) would not return a string starting with "/". Please check.
If that's the case, we could do without the second check.